apis/duck/v1/kresource_type.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"knative.dev/pkg/apis"
+)
+
+// KRShaped is an interface for retrieving the duck elements of an arbitraty resource.
+type KRShaped interface {
+	GetTypeMeta() *metav1.TypeMeta
+
+	GetObjectMeta() *metav1.ObjectMeta
+
+	GetStatus() *Status
```

Comment:
Perhaps all you need to do is revert the move of KResource and introduce two new interfaces: This KRShaped (but maybe it is called RuntimeObjectWithStatus or something) And then add an interface for the result of GetStatus: `type RuntimeStatus interface`.

This would allow you do interact will all objects that implement KResource, not just objects converted into KResource ducks.

Reply:
the duck.Status is inlined in all of the API objects (no marshalling to a different type needs to occur). The goal here is to just give common access to these fields by casting to a KRShaped.

(resolved offline over video call)